fgpyo/sam/__init__.py

```diff
@@ -1231,3 +1263,212 @@ class SamOrder(enum.Enum):
+    reference_start: int
+    is_forward: bool
+    cigar: Cigar
+    edit_distance: int
+    alignment_score: Optional[int] = None
+    mapq: Optional[int] = None
+
+    def __post_init__(self) -> None:
+        """Perform post-initialization validation on this dataclass."""
+        if self.reference_start < 0:
+            raise ValueError(f"Start cannot be <0! Found: {self.reference_start}")
+        if self.edit_distance < 0:
+            raise ValueError(f"Edit distance cannot be <0! Found: {self.edit_distance}")
+        if self.alignment_score is not None and self.alignment_score < 0:
+            raise ValueError(f"Alignment score cannot be <0! Found: {self.alignment_score}")
+        if self.mapq is not None and self.mapq < 0:
+            raise ValueError(f"Mapping quality cannot be <0! Found: {self.mapq}")
+
+    @classmethod
+    def from_tag_part(cls, part: str) -> Self:
+        """Build a secondary alignment from a single `XA` or `XB` tag part.
+
+        Args:
+            part: A single element in an `XA` or `XB` tag value.
+        """
+        fields: list[str] = part.rstrip(",").split(",")
+
+        if len(fields) == 4:
+            reference_name, stranded_start, cigar, edit_distance = fields
+            alignment_score = None
+            mapq = None
+        elif len(fields) == 6:
+            reference_name, stranded_start, cigar, edit_distance, alignment_score, mapq = fields
+        else:
+            raise ValueError(f"XA or XB tag part does not have 4 or 6 ',' separated fields: {part}")
+
+        if len(stranded_start) <= 1 or stranded_start[0] not in {"+", "-"}:
+            raise ValueError(f"The stranded start field is malformed: {stranded_start}")
+
+        return cls(
+            reference_name=reference_name,
+            reference_start=int(stranded_start[1:]) - 1,
+            is_forward=stranded_start[0] == "+",
+            cigar=Cigar.from_cigarstring(cigar),
+            edit_distance=int(edit_distance),
+            alignment_score=None if alignment_score is None else int(alignment_score),
+            mapq=None if mapq is None else int(mapq),
+        )
+
+    @classmethod
+    def many_from_tag(cls, value: str) -> list[Self]:
+        """Build many secondary alignments from a single `XA` or `XB` tag value.
+
+        Args:
+            value: A single `XA` or `XB` tag value.
+        """
+        return [cls.from_tag_part(part) for part in value.rstrip(";").split(";")]
+
+    @classmethod
+    def many_from_rec(cls, rec: AlignedSegment) -> list[Self]:
+        """Build many secondary alignments from a single SAM record.
+
+        Args:
+            rec: The SAM record to generate secondary alignments from.
+        """
+        secondaries: list[Self] = []
+        if rec.has_tag("XA"):
+            secondaries.extend(cls.many_from_tag(cast(str, rec.get_tag("XA"))))
+        if rec.has_tag("XB"):
+            secondaries.extend(cls.many_from_tag(cast(str, rec.get_tag("XB"))))
```

Comment:
**suggestion** Where it's possible to do so, I think it's helpful to include the query name when a record is malformatted

```suggestion
        if rec.has_tag("XB"):
            try:
                secondaries.extend(cls.many_from_tag(cast(str, rec.get_tag("XB"))))
            except ValueError as e:
                raise ValueError(f"Could not parse secondary alignments for read: {rec.query_name}\n{e}") from e
```